### PR TITLE
Customisable frame metrics widget

### PR DESCRIPTION
**Pull request by MutantAura**
Follow-up to #2638 
It was mentioned that this feature may be nice to have and it was very basic in nature.

New setting in "General" which is an integer spinwheel for the number of decimal places the bottom render widget will display FPS to. This defaults to the value of 2 for new configuration files. Integer rounding value scales from 0-15 decimal places.
image

See examples: 
2 decimal places
image
1 decimal place
image
0 decimal places
image
15 decimal places (perhaps quite excessive...)
image

There are some drawbacks. Using a rounding method to allow fully flexible control over this precision means if the final value is 0 then it will be truncated, cases where the user has specified 2DP and the value is 60.00 would be shortened to 60. Could possibly look into a `ToString()` method for each value but I've not thought of a way to do this efficiently yet as 15 if statements sounds messy... Might be more viable to do that if the values were restricted to 0-2 but I'm open to better ideas.
See 2DP defined but 0 has been truncated.
image

### CODE REVIEW

**Comment by EliEron** [comment]
While I approve of this PR (as I prefer less precision personally) I agree with gdkchan that a dropdown would make far more sense. As you note yourself 15 digits of precision is quite ridiculous, honestly even 3 digits of precision seems overkill. So a short dropdown like the one gdkchan proposes seems far more intuitive to me.

**Comment by MutantAura** [comment]
Updated implementation:

-Dropdown with common options instead of scroll wheel
-Fixed truncation of ending and leading 0's
-Fixes infinity frametime readings

**Review by AcK77 — CHANGES_REQUESTED** [review]

File: Ryujinx.HLE/HLEConfiguration.cs [hunk 1]
```diff
@@ -109,6 +109,11 @@ public class HLEConfiguration
         /// <remarks>This cannot be changed after <see cref="Switch"/> instantiation.</remarks>
         internal readonly int FsGlobalAccessLogMode;
 
+        /// <summary>
+        /// Integer scale to adjust frame metric displayed value. Values from 0-4
+        /// </summary>
+        public static int fpsPrecision;
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx.HLE/HLEConfiguration.cs]: Didn't need any HLE references anyway so fixed by deletion.
Unsure why I added them in the first place.

File: Ryujinx.HLE/HLEConfiguration.cs [hunk 2]
```diff
@@ -163,7 +168,8 @@ public class HLEConfiguration
                                 string                 timeZone,
                                 MemoryManagerMode      memoryManagerMode,
                                 bool                   ignoreMissingServices,
-                                AspectRatio            aspectRatio)
+                                AspectRatio            aspectRatio,
+                                int                    fpsPrecision)
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx.HLE/HLEConfiguration.cs]: As above, fixed by deletion.

File: Ryujinx.Headless.SDL2/Options.cs [hunk 3]
```diff
@@ -85,6 +85,9 @@ public class Options
         [Option("fs-global-access-log-mode", Required = false, Default = 0, HelpText = "Enables FS access log output to the console.")]
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public int fpsPrecision { get; set; }
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx.Headless.SDL2/Options.cs]: Naming has been updated globally.

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 4]
```diff
@@ -189,16 +191,62 @@ public void Render()
...
+                                frameTimeString = "";
+                                break;
+                            
+                            case 4:
+                                fpsString       = fps.ToString("00 FPS");
+                                frameTimeString = "";
+                                break;
+                        }
+
+
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx.Headless.SDL2/WindowBase.cs]: Fixed

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 5]
```diff
@@ -250,6 +257,7 @@ public SystemSection()
                 FsGlobalAccessLogMode         = new ReactiveObject<int>();
                 FsGlobalAccessLogMode.Event   += static (sender, e) => LogValueChange(sender, e, nameof(FsGlobalAccessLogMode));
                 AudioBackend                  = new ReactiveObject<AudioBackend>();
+                AudioBackend                  = new ReactiveObject<AudioBackend>();
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx/Configuration/ConfigurationState.cs]: hehe. fixed

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 6]
```diff
@@ -54,6 +54,10 @@ public abstract class RendererWidgetBase : DrawingArea
 
         private long _ticks = 0;
 
+        private string fpsString;
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx/Ui/RendererWidgetBase.cs]: string naming updated

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 7]
```diff
@@ -417,17 +421,61 @@ public void Render()
                     if (_ticks >= _ticksPerFrame)
                     {
                         string dockedMode = ConfigurationState.Instance.System.EnableDockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float scale       = Graphics.Gpu.GraphicsConfig.ResScale;
```
[the remark this answers is not in the document]
Comment (MutantAura) [reply Ryujinx/Ui/RendererWidgetBase.cs]: alignments should now be variable and value aligned

File: Ryujinx.Headless.SDL2/Options.cs [hunk 8]
```diff
@@ -85,6 +85,9 @@ public class Options
         [Option("fs-global-access-log-mode", Required = false, Default = 0, HelpText = "Enables FS access log output to the console.")]
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public static int FpsPrecision { get; set; }
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/Options.cs:89]: It shouldn't be `static`

Comment (MutantAura) [reply Ryujinx.Headless.SDL2/Options.cs]: I _think_ this is fixed...

File: Ryujinx.Headless.SDL2/OpenGL/OpenGLWindow.cs [hunk 9]
```diff
@@ -103,7 +103,7 @@ public void Dispose()
         private GraphicsDebugLevel _glLogLevel;
         private SDL2OpenGLContext _openGLContext;
 
-        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse) : base(inputManager, glLogLevel, aspectRatio, enableMouse)
+        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, int FpsPrecision) : base(inputManager, glLogLevel, aspectRatio, enableMouse, FpsPrecision)
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/OpenGL/OpenGLWindow.cs:106]:
```suggestion
        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, uint fpsPrecision) : base(inputManager, glLogLevel, aspectRatio, enableMouse, fpsPrecision)
```

File: Ryujinx.Headless.SDL2/Options.cs [hunk 10]
```diff
@@ -85,6 +85,9 @@ public class Options
         [Option("fs-global-access-log-mode", Required = false, Default = 0, HelpText = "Enables FS access log output to the console.")]
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public int FpsPrecision { get; set; }
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/Options.cs:89]:
```suggestion
        public uint FpsPrecision { get; set; }
```

The value is between 0 and 4 so a uint should be enough

File: Ryujinx.Headless.SDL2/Program.cs [hunk 11]
```diff
@@ -500,7 +500,7 @@ private static bool LoadApplication(Options options)
 
             Logger.RestartTime();
 
-            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse);
+            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse, (int)options.FpsPrecision);
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/Program.cs:503]:
```suggestion
            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse, options.FpsPrecision);
```

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 12]
```diff
@@ -52,8 +54,9 @@ abstract class WindowBase : IHostUiHandler, IDisposable
 
         private AspectRatio _aspectRatio;
         private bool _enableMouse;
+        private int fpsPrecision;
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/WindowBase.cs:57]:
```suggestion
        private uint _fpsPrecision;
```

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 13]
```diff
@@ -52,8 +54,9 @@ abstract class WindowBase : IHostUiHandler, IDisposable
...
 
-        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse)
+        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, int FpsPrecision)
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/WindowBase.cs:59]:
```suggestion
        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, uint fpsPrecision)
```

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 14]
```diff
@@ -67,6 +70,7 @@ public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, Aspe
             _exitEvent = new ManualResetEvent(false);
             _aspectRatio = aspectRatio;
             _enableMouse = enableMouse;
+            fpsPrecision = FpsPrecision;
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/WindowBase.cs:73]:
```suggestion
            _fpsPrecision = fpsPrecision;
```

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 15]
```diff
@@ -188,17 +192,61 @@ public void Render()
                     if (_ticks >= _ticksPerFrame)
                     {
                         string dockedMode = Device.System.State.DockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float  scale      = Graphics.Gpu.GraphicsConfig.ResScale;
+                        int    precision  = fpsPrecision;
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/WindowBase.cs:196]: You can remove this line

File: Ryujinx.Headless.SDL2/WindowBase.cs [hunk 16]
```diff
@@ -188,17 +192,61 @@ public void Render()
...
+                        double fps        = Device.Statistics.GetGameFrameRate();
+                        double frameTime  = Device.Statistics.GetGameFrameTime();
+
                         if (scale != 1)
                         {
                             dockedMode += $" ({scale}x)";
                         }
 
+                        switch (precision)
```
Comment (AcK77) [inline Ryujinx.Headless.SDL2/WindowBase.cs:205]:
```suggestion
                        switch (_fpsPrecision)
```

File: Ryujinx/Configuration/ConfigurationFileFormat.cs [hunk 17]
```diff
@@ -174,6 +174,11 @@ public class ConfigurationFileFormat
         /// </summary>
         public int FsGlobalAccessLogMode { get; set; }
 
+        /// <summary>
+        /// Integer scale to adjust frame metric displayed value. Values from 0-4
+        /// </summary>
+        public int FpsPrecision { get; set; }
```
Comment (AcK77) [inline Ryujinx/Configuration/ConfigurationFileFormat.cs:180]:
```suggestion
        public uint FpsPrecision { get; set; }
```

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 18]
```diff
@@ -88,14 +88,21 @@ public ColumnSortSettings()
             /// </summary>
             public ReactiveObject<bool> StartFullscreen { get; private set; }
 
+            // <summary>
+            /// Integer scale to adjust frame metric displayed value. Values from 0-4
+            /// </summary>
+            public ReactiveObject<int> FpsPrecision { get; private set; }
```
Comment (AcK77) [inline Ryujinx/Configuration/ConfigurationState.cs:94]:
```suggestion
            public ReactiveObject<uint> FpsPrecision { get; private set; }
```

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 19]
```diff
@@ -88,14 +88,21 @@ public ColumnSortSettings()
...
-                EnableCustomTheme = new ReactiveObject<bool>();
-                CustomThemePath   = new ReactiveObject<string>();
-                StartFullscreen   = new ReactiveObject<bool>();
+                GuiColumns         = new Columns();
+                ColumnSort         = new ColumnSortSettings();
+                GameDirs           = new ReactiveObject<List<string>>();
+                EnableCustomTheme  = new ReactiveObject<bool>();
+                CustomThemePath    = new ReactiveObject<string>();
+                StartFullscreen    = new ReactiveObject<bool>();
+                FpsPrecision       = new ReactiveObject<int>();
```
Comment (AcK77) [inline Ryujinx/Configuration/ConfigurationState.cs:104]:
```suggestion
                FpsPrecision       = new ReactiveObject<uint>();
```

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 20]
```diff
@@ -54,6 +54,10 @@ public abstract class RendererWidgetBase : DrawingArea
 
         private long _ticks = 0;
 
+        private string _fpsString;
+
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:58]: Remove the extra empty line

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 21]
```diff
@@ -417,17 +421,61 @@ public void Render()
                     if (_ticks >= _ticksPerFrame)
                     {
                         string dockedMode = ConfigurationState.Instance.System.EnableDockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float  scale      = Graphics.Gpu.GraphicsConfig.ResScale;
+                        int    precision  = ConfigurationState.Instance.Ui.FpsPrecision;
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:425]: You can remove this line

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 22]
```diff
@@ -417,17 +421,61 @@ public void Render()
...
+                        double fps        = Device.Statistics.GetGameFrameRate();
+                        double frameTime  = Device.Statistics.GetGameFrameTime();
+                        
                         if (scale != 1)
                         {
                             dockedMode += $" ({scale}x)";
                         }
 
+                        switch (precision)
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:434]:
```suggestion
                        switch (ConfigurationState.Instance.Ui.FpsPrecision)
```

File: Ryujinx/Ui/Windows/SettingsWindow.cs [hunk 23]
```diff
@@ -305,6 +306,7 @@ private SettingsWindow(MainWindow parent, Builder builder, VirtualFileSystem vir
             _resScaleCombo.SetActiveId(ConfigurationState.Instance.Graphics.ResScale.Value.ToString());
             _anisotropy.SetActiveId(ConfigurationState.Instance.Graphics.MaxAnisotropy.Value.ToString());
             _aspectRatio.SetActiveId(((int)ConfigurationState.Instance.Graphics.AspectRatio.Value).ToString());
+            _fpsPrecisionCombo.SetActiveId(ConfigurationState.Instance.Ui.FpsPrecision.Value.ToString());
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/SettingsWindow.cs:309]:
```suggestion
            _fpsPrecisionCombo.SetActiveId(((int)ConfigurationState.Instance.Ui.FpsPrecision.Value).ToString());
```

File: Ryujinx/Ui/Windows/SettingsWindow.cs [hunk 24]
```diff
@@ -493,6 +495,7 @@ private void SaveSettings()
             ConfigurationState.Instance.Graphics.ShadersDumpPath.Value         = _graphicsShadersDumpPath.Buffer.Text;
             ConfigurationState.Instance.Ui.GameDirs.Value                      = gameDirs;
             ConfigurationState.Instance.System.FsGlobalAccessLogMode.Value     = (int)_fsLogSpinAdjustment.Value;
+            ConfigurationState.Instance.Ui.FpsPrecision.Value                  = int.Parse(_fpsPrecisionCombo.ActiveId);
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/SettingsWindow.cs:498]:
```suggestion
            ConfigurationState.Instance.Ui.FpsPrecision.Value                  = uint.Parse(_fpsPrecisionCombo.ActiveId);
```

**Review by AcK77 — CHANGES_REQUESTED** [review]

**Review by AcK77 — APPROVED** [review]
Lgtm! Thanks!

**Pull request closed without merging**